cmd/snap-seccomp/main.go

```diff
@@ -0,0 +1,522 @@
+	ScmpArchPPC64LE = C.SCMP_ARCH_PPC64LE
+	ScmpArchS390X   = C.SCMP_ARCH_S390X
+
+	SeccompRetAllow = C.SECCOMP_RET_ALLOW
+	SeccompRetKill  = C.SECCOMP_RET_KILL
+)
+
+// important for unit testing
+type SeccompData C.kernel_seccomp_data
+
+func (sc *SeccompData) SetNr(nr seccomp.ScmpSyscall) {
+	sc.nr = C.int(C.htot32(C.__u32(sc.arch), C.__u32(nr)))
+}
+func (sc *SeccompData) SetArch(arch uint32) {
+	sc.arch = C.htot32(C.__u32(arch), C.__u32(arch))
+}
+func (sc *SeccompData) SetArgs(args [6]uint64) {
+	for i := range args {
+		sc.args[i] = C.htot64(sc.arch, C.__u64(args[i]))
+	}
+}
+
+func readNumber(token string) (uint64, error) {
+	if value, ok := seccompResolver[token]; ok {
+		return value, nil
+	}
+
+	return strconv.ParseUint(token, 10, 64)
+}
+
+func parseLine(line string, secFilter *seccomp.ScmpFilter) error {
+	line = strings.TrimSpace(line)
+
+	// ignore comments and empty lines
+	if strings.HasPrefix(line, "#") || line == "" {
+		return nil
+	}
+
+	// regular line
+	tokens := strings.Fields(line)
+
+	// fish out syscall
+	secSyscall, err := seccomp.GetSyscallFromName(tokens[0])
+	if err != nil {
+		// FIXME: add structed error to libseccomp-golang
+		if err.Error() == "could not resolve name to syscall" {
+			return nil
+		}
+		return fmt.Errorf("cannot resolve name: %s", err)
+	}
+
+	var conds []seccomp.ScmpCondition
+	if err != nil {
+		return fmt.Errorf("cannot create new filter: %s", err)
+	}
+
+	for pos, arg := range tokens[1:] {
+		var cmpOp seccomp.ScmpCompareOp
+		var value uint64
+		var err error
+
+		if pos >= ScArgsMaxlength {
+			return fmt.Errorf("too many tokens (%d) in line %q", pos, line)
+		}
+
+		if arg == "-" {
+			continue
+		}
+
+		if strings.HasPrefix(arg, ">=") {
+			cmpOp = seccomp.CompareGreaterEqual
+			value, err = readNumber(arg[2:])
+		} else if strings.HasPrefix(arg, "<=") {
+			cmpOp = seccomp.CompareLessOrEqual
+			value, err = readNumber(arg[2:])
+		} else if strings.HasPrefix(arg, "!") {
+			cmpOp = seccomp.CompareNotEqual
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, "<") {
+			cmpOp = seccomp.CompareLess
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, ">") {
+			cmpOp = seccomp.CompareGreater
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, "|") {
+			cmpOp = seccomp.CompareMaskedEqual
+			value, err = readNumber(arg[1:])
+		} else {
+			cmpOp = seccomp.CompareEqual
+			value, err = readNumber(arg)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse token %q (line %q)", arg, line)
+		}
+
+		var scmpCond seccomp.ScmpCondition
+		if cmpOp == seccomp.CompareMaskedEqual {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value, value)
+		} else {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse line %q: %s", line, err)
+		}
+		conds = append(conds, scmpCond)
+	}
+
+	// FIXME: why do we need this fallback?
```

Comment:
seccomp_rule_add_exact_array is more precise than seccomp_rule_add_array (see `man seccomp_rule_add`). I suggest adding a comment:
```
// Default to adding a precise match if possible. Otherwise let seccomp figure
// out the architecture specifics.
```